### PR TITLE
Hott-1284 Add Slack notification.

### PR DESCRIPTION
**Pull request by alexdesi**
### Jira link

[HOTT-<1284](https://transformuk.atlassian.net/browse/HOTT-1284)

### What?

Add Slack notification in case of FailedUpdatesError

### CODE REVIEW

File: app/services/slack_notifier_service.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        return
+      end
+
+      notifier.ping(message)
+    end
+
+  private
+
+    def notifier
+      @msg_notifier || slack_notifier
```
Comment (alexdesi) [inline app/services/slack_notifier_service.rb:21]: I am injecting the  @msg_notifier to simplify tests, even though I could probably use allowAnyInstanceOf (which is deprecated anyway)

File: app/services/slack_notifier_service.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
+class SlackNotifierService
```
Comment (jebw) [inline app/services/slack_notifier_service.rb:1]: This is very much personal opinion but I'd suggest rather then having this as a static class, ie

```
SlackNotifierService.call
```

instead differentiate between 'action' data and maintained state data, eg

```
SlackNotifierService.new(msg_notifier).notify('my message to slack')
```

File: app/services/slack_notifier_service.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  CHANNEL = '#tariffs-etl'.freeze
+  USERNAME = 'Trade Tariff BE'.freeze
```
Comment (jebw) [inline app/services/slack_notifier_service.rb:3]: What is BE? If this is backend, can it be changed to `Trade Tariff Backend` - that would remove the risk of a future developer not knowing what BE means at the cost of 5 letters.

**Review by jebw — DISMISSED** [review]
Looks good to me - only comments are a code style suggestion and the abbreviation needs to be replaced.

File: app/lib/tariff_synchronizer.rb [hunk 4]
```diff
@@ -324,5 +328,13 @@ def check_tariff_updates_failures
...
       raise FailedUpdatesError
     end
+  rescue FailedUpdatesError => e
+    notify_slack_app(e)
+
+    raise
+  end
+
+  def notify_slack_app(exception)
+    SlackNotifierService.call("Error #{exception.class}: #{exception.message}")
```
Comment (willfish) [inline app/lib/tariff_synchronizer.rb:338]: I don't think this passes the signature requirements of the call method?

**Review by jebw — CHANGES_REQUESTED** [review]
Looks good to me - only comments are a code style suggestion and the abbreviation needs to be replaced.

Marked as needs changing because the PRs still got failing specs?

File: app/lib/tariff_synchronizer.rb [hunk 5]
```diff
@@ -324,5 +328,13 @@ def check_tariff_updates_failures
                  file_names: BaseUpdate.failed.map(&:filename))
       raise FailedUpdatesError
     end
+  rescue FailedUpdatesError => e
+    notify_slack_app(e)
+
+    raise
+  end
+
+  def notify_slack_app(exception)
```
Comment (willfish) [inline app/lib/tariff_synchronizer.rb:337]: Should we add a test for this integration with the service? I suspect this isn't actually passing?

File: app/services/slack_notifier_service.rb (hunk 1 above)
Comment (willfish) [reply app/services/slack_notifier_service.rb:21]: I wonder if it would be better to configure slack notifier as part of an initialiser for this gem?

Comment (willfish) [reply app/services/slack_notifier_service.rb:21]: That also feels like a better place to access environment variables

File: app/lib/tariff_synchronizer.rb (hunk 5 above)
Comment (willfish) [reply app/lib/tariff_synchronizer.rb:337]: Just a light touch one:

allow(slack_notifier_service).to receive(:call)
expect(slack_notifier_service).to have_receive(:call).with(expected_message)

File: app/services/slack_notifier_service.rb (hunk 2 above)
Comment (alexdesi) [reply app/services/slack_notifier_service.rb:1]: it sounds good to me, my choice was influenced by the way we do services. I'll give it a try.

File: config/initializers/slack_notifier.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+require 'slack-notifier'
+
+if ENV['SLACK_WEB_HOOK_URL'] && ENV['SLACK_CHANNEL'] && ENV['SLACK_USERNAME']
```
Comment (willfish) [inline config/initializers/slack_notifier.rb:3]: Maybe just add handling for production only env and for the tests you can construct a specific double/mock/stub in the spec helper?

return unless Rails.env.production?

File: spec/services/slack_notifier_service_spec.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+RSpec.describe SlackNotifierService do
+  before do
+    allow(Rails.application.config.slack_notifier).to receive(:ping)
```
Comment (willfish) [inline spec/services/slack_notifier_service_spec.rb:3]: Perfect :)

File: config/initializers/slack_notifier.rb (hunk 6 above)
Comment (alexdesi) [reply config/initializers/slack_notifier.rb:3]: not sure, in that case we should probably move that to the `/environments/production.rb`

Comment (willfish) [reply config/initializers/slack_notifier.rb:3]: This is the same approach you've taken elsewhere, though https://github.com/trade-tariff/trade-tariff-backend/commit/052c310553c107a7ebf5381e0532ac3c1f487df0

Comment (willfish) [reply config/initializers/slack_notifier.rb:3]: I think when you're configuring a specific gem regardless of how we want it behave the approach you've taken elsewhere is the correct one.

**Review by willfish — APPROVED** [review]

**Comment by willfish** [comment]
A couple of the lint violations are to do  with the files we've added :smile:

**Review by willfish — APPROVED** [review]

File: app/services/slack_notifier_service.rb (hunk 2 above)
Comment (alexdesi) [reply app/services/slack_notifier_service.rb:1]: I made some changes, now the notifier is stored in the Rails.application.config.slack_notifier.

**Pull request merged**